components/ws-manager-bridge/src/bridge.ts

```diff
@@ -503,6 +498,75 @@ export class WorkspaceManagerBridge implements Disposable {
         }
     }
 
+    /**
+     * This methods controls all instances of this installation during periods where ws-manager does not control them, but we have them in our DB.
+     * These currently are:
+     *  - preparing
+     *  - building
+     * It also covers these phases, as fallback, when - for whatever reason - we no longer receive updates from ws-manager.
+     *  - unknown (fallback)
+     */
+    protected async controlInstancesTimeouts(parentCtx: TraceContext, runningInstances: RunningWorkspaceInfo[]) {
+        const installation = this.config.installation;
+
+        const span = TraceContext.startSpan("controlDBInstances", parentCtx);
+        const ctx = { span };
+        try {
+            log.debug("Controlling DB instances...", { installation });
+
+            await Promise.all(runningInstances.map((info) => this.controlInstanceTimeouts(ctx, info)));
+
+            log.debug("Done controlling DB instances.", { installation });
+        } catch (err) {
+            log.error("Error while running controlDBInstances", err, {
+                installation: this.cluster.name,
+            });
+            TraceContext.setError(ctx, err);
+        } finally {
+            span.finish();
+        }
+    }
+
+    protected async controlInstanceTimeouts(parentCtx: TraceContext, info: RunningWorkspaceInfo) {
+        const logContext: LogContext = {
+            userId: info.workspace.ownerId,
+            workspaceId: info.workspace.id,
+            instanceId: info.latestInstance.id,
+        };
+        const ctx = TraceContext.childContext("controlDBInstance", parentCtx);
+        try {
+            const now = Date.now();
+            const creationTime = new Date(info.latestInstance.creationTime).getTime();
+            const timedOutInPreparing = now >= creationTime + this.config.timeouts.preparingPhaseSeconds * 1000;
+            const timedOutInBuilding = now >= creationTime + this.config.timeouts.buildingPhaseSeconds * 1000;
+            const timedOutInUnknown = now >= creationTime + this.config.timeouts.unknownPhaseSeconds * 1000;
+            const currentPhase = info.latestInstance.status.phase;
```

Comment:
It is somewhat annoying that we have to do this here. Ideally, we'd always be able to ask a ws-manager about a particular WSI and it would tell us if it is in fact in a failed state based on these timeouts.